src/Access/IAccessStorage.cpp

```diff
@@ -531,9 +539,31 @@ std::optional<UUID> IAccessStorage::authenticateImpl(
                 ((auth_type == AuthenticationType::PLAINTEXT_PASSWORD) && !allow_plaintext_password))
                 throwAuthenticationTypeNotAllowed(auth_type);
 
+            auto new_current_roles = user->granted_roles.findGranted(user->default_roles);
+            const auto & access = Context::getGlobalContextInstance()->getAccessControl();
+            auto roles_info = access.getEnabledRolesInfo(new_current_roles, {});
+            const std::string custom_quota_key = ""; // TODO: Where do we get it?
+
+            assert(assert);
+            auto enabled_quota = access.getEnabledQuota(*id,
+                                                        credentials.getUserName(),
+                                                        roles_info->enabled_roles,
+                                                        address,
+                                                        forwarded_address,
```

Comment:
How can I get a quota key from the TCP client here?

At the moment of the authentication, we have not received client_key (quota_key) from the client.

Reply:
I created a function getting quota for authentication with `constant` quota key :

```cpp
    std::shared_ptr<const EnabledQuota> getAuthenticationQuota(UUID user_id,
                                                               const UserPtr & user,
                                                               const Poco::Net::IPAddress & address,
                                                               const std::string & forwarded_address)
    {
        /// During authentication process, client_key is not received from TCP client,
        /// Use predefined authentication quota key and always receive the save Interval object
        /// to avoid throwing exceptions in case of QuotaKeyType::CLIENT_KEY key type.
        constexpr auto AUTHENTICATION_QUOTA_KEY = "_AUTHENTICATION_QUOTA_KEY_";
        auto new_current_roles = user->granted_roles.findGranted(user->default_roles);
        const auto & access = Context::getGlobalContextInstance()->getAccessControl();
        auto roles_info = access.getEnabledRolesInfo(new_current_roles, {});

        auto enabled_quota = access.getEnabledQuota(user_id,
                                                    user->getName(),
                                                    roles_info->enabled_roles,
                                                    address,
                                                    forwarded_address,
                                                    AUTHENTICATION_QUOTA_KEY);

        return enabled_quota;
    }

```

This solution will ignore the futher quota received by TCP connection. This solution will work with `QuotaKeyType::CLIENT_KEY`.

Reply:
`_AUTHENTICATION_QUOTA_KEY_` it looks too scary; we shouldn't have it in the code...

Reply:
Good day, Alexey

I updated the code. Now it does not have conflicts and `_AUTHENTICATION_QUOTA_KEY_`,
but still we have to provide quota_key to the `AccessControl::getEnabledQuota`, during TCP connection, quota_key (i.e client_key) is sent after authentication, but failed login attempt count should be checked before authentication is finished, so we can't get quota key from the user.

Without anything passed to the quota, we will get an exception in case of `QuotaKeyType::CLIENT_KEY` from this code:

```cpp
String QuotaCache::QuotaInfo::calculateKey(const EnabledQuota & enabled) const
{
...
        throw Exception(
            ErrorCodes::QUOTA_REQUIRES_CLIENT_KEY,
            "Quota {} (for user {}) requires a client supplied key.",
            quota->getName(),
            params.user_name);
...
}
```

As a workaround, I added a parameter `throw_if_client_key_empty` for the authentication quota case.
```cpp
String QuotaCache::QuotaInfo::calculateKey(const EnabledQuota & enabled, bool throw_if_client_key_empty) const
{
...
   if (throw_if_client_key_empty)
        throw Exception(
            ErrorCodes::QUOTA_REQUIRES_CLIENT_KEY,
            "Quota {} (for user {}) requires a client supplied key.",
            quota->getName(),
            params.user_name);
...
}
```

The issue is solved, but I'm not sure that this solution is good enough.

Please advise me if you have a better idea of how to solve this.